diff -Naur sed-4.6-orig/lib/ctype.h sed-4.6-patched/lib/ctype.h
--- sed-4.6-orig/lib/ctype.h	2018-12-21 13:12:20.551404400 +0300
+++ sed-4.6-patched/lib/ctype.h	2018-12-21 12:56:00.638032400 +0300
@@ -32,7 +32,7 @@
 
 /* Include the original <ctype.h>.  */
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\ctype.h"
 
 #ifndef _GL_CTYPE_H
 #define _GL_CTYPE_H
diff -Naur sed-4.6-orig/lib/errno.h sed-4.6-patched/lib/errno.h
--- sed-4.6-orig/lib/errno.h	2018-12-21 13:12:20.701404600 +0300
+++ sed-4.6-patched/lib/errno.h	2018-12-21 12:56:11.558047700 +0300
@@ -24,7 +24,7 @@
 
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\errno.h"
 
 #ifndef _GL_ERRNO_H
 #define _GL_ERRNO_H
diff -Naur sed-4.6-orig/lib/fcntl.h sed-4.6-patched/lib/fcntl.h
--- sed-4.6-orig/lib/fcntl.h	2018-12-21 13:12:20.881404900 +0300
+++ sed-4.6-patched/lib/fcntl.h	2018-12-21 12:56:37.418083900 +0300
@@ -58,7 +58,7 @@
 # include <sys/stat.h>
 #endif
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\fcntl.h"
 
 #ifndef _GL_FCNTL_H
 #define _GL_FCNTL_H
diff -Naur sed-4.6-orig/lib/limits.h sed-4.6-patched/lib/limits.h
--- sed-4.6-orig/lib/limits.h	2018-12-21 13:12:21.421405700 +0300
+++ sed-4.6-patched/lib/limits.h	2018-12-21 12:56:50.228101900 +0300
@@ -24,7 +24,7 @@
 
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\limits.h"
 
 #ifndef _GL_LIMITS_H
 #define _GL_LIMITS_H
diff -Naur sed-4.6-orig/lib/locale.h sed-4.6-patched/lib/locale.h
--- sed-4.6-orig/lib/locale.h	2018-12-21 13:12:21.651406000 +0300
+++ sed-4.6-patched/lib/locale.h	2018-12-21 12:57:20.308144000 +0300
@@ -38,7 +38,7 @@
 #define _GL_ALREADY_INCLUDING_LOCALE_H
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\locale.h"
 
 #undef _GL_ALREADY_INCLUDING_LOCALE_H
 
diff -Naur sed-4.6-orig/lib/stddef.h sed-4.6-patched/lib/stddef.h
--- sed-4.6-orig/lib/stddef.h	2018-12-21 13:12:22.161406700 +0300
+++ sed-4.6-patched/lib/stddef.h	2018-12-21 12:57:40.668172500 +0300
@@ -52,7 +52,7 @@
 
 /* The include_next requires a split double-inclusion guard.  */
 
-#  include ""
+#  include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\stddef.h"
 
 /* On NetBSD 5.0, the definition of NULL lacks proper parentheses.  */
 #  if (0 \
diff -Naur sed-4.6-orig/lib/stdint.h sed-4.6-patched/lib/stdint.h
--- sed-4.6-orig/lib/stdint.h	2018-12-21 13:12:22.401407000 +0300
+++ sed-4.6-patched/lib/stdint.h	2018-12-21 12:58:03.328204200 +0300
@@ -74,7 +74,7 @@
      in <inttypes.h> would reinclude us, skipping our contents because
      _GL_STDINT_H is defined.
      The include_next requires a split double-inclusion guard.  */
-# include ""
+# include "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\stdint.h"
 #endif
 
 #if ! defined _GL_STDINT_H && ! defined _GL_JUST_INCLUDE_SYSTEM_STDINT_H
diff -Naur sed-4.6-orig/lib/stdio.h sed-4.6-patched/lib/stdio.h
--- sed-4.6-orig/lib/stdio.h	2018-12-21 13:12:23.031407900 +0300
+++ sed-4.6-patched/lib/stdio.h	2018-12-21 12:58:19.768227200 +0300
@@ -40,7 +40,7 @@
 #define _GL_ALREADY_INCLUDING_STDIO_H
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\stdio.h"
 
 #undef _GL_ALREADY_INCLUDING_STDIO_H
 
diff -Naur sed-4.6-orig/lib/stdlib.h sed-4.6-patched/lib/stdlib.h
--- sed-4.6-orig/lib/stdlib.h	2018-12-21 13:12:23.571408700 +0300
+++ sed-4.6-patched/lib/stdlib.h	2018-12-21 12:59:31.848328100 +0300
@@ -33,7 +33,7 @@
 #ifndef _GL_STDLIB_H
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\stdlib.h"
 
 #ifndef _GL_STDLIB_H
 #define _GL_STDLIB_H
diff -Naur sed-4.6-orig/lib/string.h sed-4.6-patched/lib/string.h
--- sed-4.6-orig/lib/string.h	2018-12-21 13:12:24.091409400 +0300
+++ sed-4.6-patched/lib/string.h	2018-12-21 12:59:45.798347700 +0300
@@ -38,7 +38,7 @@
 #define _GL_ALREADY_INCLUDING_STRING_H
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\string.h"
 
 #undef _GL_ALREADY_INCLUDING_STRING_H
 
diff -Naur sed-4.6-orig/lib/sys/stat.h sed-4.6-patched/lib/sys/stat.h
--- sed-4.6-orig/lib/sys/stat.h	2018-12-21 13:12:24.431409900 +0300
+++ sed-4.6-patched/lib/sys/stat.h	2018-12-21 13:00:30.868410800 +0300
@@ -44,7 +44,7 @@
 #include <time.h>
 
 /* The include_next requires a split double-inclusion guard.  */
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\sys\stat.h"
 
 #ifndef _GL_SYS_STAT_H
 #define _GL_SYS_STAT_H
diff -Naur sed-4.6-orig/lib/sys/types.h sed-4.6-patched/lib/sys/types.h
--- sed-4.6-orig/lib/sys/types.h	2018-12-21 13:12:24.771410300 +0300
+++ sed-4.6-patched/lib/sys/types.h	2018-12-21 13:00:53.538442500 +0300
@@ -36,7 +36,7 @@
 
 /* The include_next requires a split double-inclusion guard.  */
 # define _GL_INCLUDING_SYS_TYPES_H
-#include ""
+#include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\sys\types.h"
 # undef _GL_INCLUDING_SYS_TYPES_H
 
 #ifndef _GL_SYS_TYPES_H
diff -Naur sed-4.6-orig/lib/time.h sed-4.6-patched/lib/time.h
--- sed-4.6-orig/lib/time.h	2018-12-21 13:12:25.041410700 +0300
+++ sed-4.6-patched/lib/time.h	2018-12-21 13:01:35.688501500 +0300
@@ -38,7 +38,7 @@
 
 # define _GL_TIME_H
 
-# include ""
+# include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\time.h"
 
 /* NetBSD 5.0 mis-defines NULL.  */
 # include <stddef.h>
diff -Naur sed-4.6-orig/lib/wchar.h sed-4.6-patched/lib/wchar.h
--- sed-4.6-orig/lib/wchar.h	2018-12-21 13:12:26.511412800 +0300
+++ sed-4.6-patched/lib/wchar.h	2018-12-21 13:02:01.648537900 +0300
@@ -84,7 +84,7 @@
    Some builds of uClibc lack it.  */
 /* The include_next requires a split double-inclusion guard.  */
 #if 1
-# include ""
+# include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\wchar.h"
 #endif
 
 #undef _GL_ALREADY_INCLUDING_WCHAR_H
diff -Naur sed-4.6-orig/lib/wctype.h sed-4.6-patched/lib/wctype.h
--- sed-4.6-orig/lib/wctype.h	2018-12-21 13:12:26.721413100 +0300
+++ sed-4.6-patched/lib/wctype.h	2018-12-21 13:02:13.228554100 +0300
@@ -70,7 +70,7 @@
    BeOS 5 has the functions but no <wctype.h>.  */
 /* The include_next requires a split double-inclusion guard.  */
 #if 1
-# include ""
+# include "C:\Program Files (x86)\Windows Kits\10\Include\10.0.14393.0\ucrt\wctype.h"
 #endif
 
 #ifndef _GL_WCTYPE_H
